sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1162 @@
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `cgf` - Command group function object to be added as a node.
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The command-group function object node which has been added to the graph.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+void make_edge(node src, node dest);
+----
+
+|Creates a dependency between two nodes representing a happens-before relationship.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `src` - Node which will be a dependency of `dest`.
+
+* `dest` - Node which will be dependent on `src`.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph object.
+
+* Throws synchronously with error code `invalid` if `src` or `dest`
+  are not valid nodes assigned to the graph object.
+
+* Throws synchronously with error code `invalid` if `src` and `dest`
+  are the same node.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+command_graph<graph_state::executable> finalize(const context &syclContext) const;
+----
+
+|Synchronous operation that creates a graph in the executable state with a
+fixed topology that can be submitted for execution on any queue sharing the
+supplied context. It is valid to call this method multiple times to create
+subsequent executable graphs. It is also valid to continue to add new nodes to
+the modifiable graph instance after calling this function. It is valid to
+finalize an empty graph instance with no recorded commands.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `syclContext` - The context associated with the queues to which the
+  executable graph will be able to be submitted.
+
+Returns: An executable graph object which can be submitted to a queue.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the graph contains a cycle.
+  A cycle may be introduced to the graph via a call to `make_edge()` that
+  creates a forward dependency.
+
+|===
+
+Memory that is allocated by the following functions is owned by the specific
+graph. When freed inside the graph, the memory is only accessible before the
+`free` node is executed and after the `malloc` node is executed.
+
+Table 7. Member functions of the `command_graph` class (memory operations).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add_malloc_device(void *&data, size_t numBytes, const property_list& propList = {});
```

Comment:
This (and `add_free`) look weird to me, because it really feels like there should be command-group functions for them.  It would be nicer if `malloc_device` and `free` were defined inside of `handler` so that it was possible just to treat these as regular command-groups.

I appreciate this isn't really your problem, and you probably don't want to modify `handler` more than you have to.  But it might be worth adding a non-normative note here to say that it would be nice to explore alternative designs here.

Also tagging @jbrodman, since I think this could be a useful extension, even outside of graphs.

Reply:
Yes agreed. As is, these functions are aligned with the existing free standing USM functions in SYCL. The alternative design you are describing could be beneficial, not only for the command graph extensions. I'll add a note.